src/test/java/ibm/jceplus/junit/base/BaseTestAESCopySafe.java

```diff
@@ -0,0 +1,182 @@
+        super(providerName);
+        this.specifiedKeySize = keySize;
+        Assume.assumeTrue(javax.crypto.Cipher.getMaxAllowedKeyLength("AES") >= keySize);
+    }
+
+    public void testOverlappingBuffer() throws Exception {
+
+        KEY = new SecretKeySpec(new byte[specifiedKeySize / 8], "AES");
+
+        AlgorithmParameterSpec params = null;
+        for (MODE mode : MODE.values()) {
+            String transformation = "AES/" + mode.toString() + "/NoPadding";
+            Cipher c = Cipher.getInstance(transformation, providerName);
+            System.out.println("Testing " + transformation + " from provider: " + providerName);
+            for (int inputOffset : OFFSETS) {
+                for (int outputOffset : OFFSETS) {
+                    System.out.println("Mode: " + mode + " inputOffset: " + inputOffset + " outputOffset: " + outputOffset);
+                    System.out.println("Testing doFinal");
+                    doTest(c, params, inputOffset, outputOffset, mode, false);
+                    if (mode == MODE.CBC) {
+                        System.out.println("Testing update");
+                        doTest(c, params, inputOffset, outputOffset, mode, true);
+                    }
+                }
+            }
+        }
+    }
+
+    private void doTest(Cipher c, AlgorithmParameterSpec params, int inputOffset, int outputOffset, MODE mode, boolean isUpdate)
+            throws Exception {
+        byte[] clearText = new byte[INPUT_LENGTH];
+        Arrays.fill(clearText, 0, INPUT_LENGTH, (byte)0x01);
+        System.arraycopy(clearText, 0, workingBuffer, 0, INPUT_LENGTH);
+
+        // Get baseline encrypted value. This baseline will be used through the rest of the 
+        // test to be a known answer for cipher text.
+        initCipher(c, mode, true);
+        if (DEBUG) {
+            System.out.println("Calling c." + (isUpdate ? "update" : "doFinal") + "(workingBuffer, 0, INPUT_LENGTH)");
+            System.out.println("    INPUT_LENGTH: " + INPUT_LENGTH);
+            System.out.println("    workingBuffer:\n" + BaseUtils.bytesToHex(workingBuffer));
+        }
+        byte[] cipherText = null;
+        if (isUpdate) {
+            cipherText = c.update(workingBuffer, 0, INPUT_LENGTH);
+        } else {
+            cipherText = c.doFinal(workingBuffer, 0, INPUT_LENGTH);
+        }
+        if (DEBUG) {
+            System.out.println("cipherText:\n" + BaseUtils.bytesToHex(cipherText));
+            System.out.println ("cipherText.length: " + cipherText.length);
+        }
+
+        // A GCM cipher must be reinitialized since the IV is typically not supposed to be reused. This
+        // test however expects a deterministic encrypted value in order to evaluate if the operation
+        // worked as expected.
+        initCipher(c, mode, true);
+
+        System.arraycopy(clearText, 0, workingBuffer, inputOffset, INPUT_LENGTH);
+        if (DEBUG) {
```

Comment:
Should we specify that we are testing encryption in this step?

Reply:
Yes its prob best to state we are doing encryption or decryption in this DEBUG statement.  Adjusted this in new commit https://github.com/IBM/OpenJCEPlus/pull/295/commits/46a71f1f5e9d53cc94e3336b27fbf231deaeeab3 .